src/melee/it/item.c

```diff
@@ -488,45 +448,35 @@ extern struct ItemLogicTable lbl_803F23CC[];
 
 extern struct ItemLogicTable lbl_803F4D20[];
 
-// 0x80267978 //
-// https://decomp.me/scratch/LnIJJ //
-void func_80267978(HSD_GObj* item_gobj) // Store Item Pointers //
+// Store Item Pointers
+void func_80267978(HSD_GObj* item_gobj)
 {
-    s32 ItemID;
-    s32 ItemID2;
-    unsigned int new_var;
-    Item* item_data;
-    item_data = item_gobj->user_data;
-    ItemID = item_data->x10_item_kind;
-    if (ItemID < It_Kind_Kuriboh) {
-        item_data->xC4_article_data = &(*lbl_804D6D24->article_ptr[ItemID]);
+    Item* item_data = item_gobj->user_data;
+    if (item_data->x10_item_kind < It_Kind_Kuriboh) {
+        // Common items
+        item_data->xC4_article_data =
+            lbl_804D6D24->article_ptr[item_data->x10_item_kind];
         item_data->xB8_itemLogicTable = &lbl_803F14C4[item_data->x10_item_kind];
-        goto block_end;
-    }
-    if (ItemID < Pokemon_Tosakinto) {
-        ItemID -= It_Kind_Kuriboh;
-        new_var = ItemID;
-        ItemID2 = new_var;
-        item_data->xC4_article_data = &(*lbl_804D6D38->unkptr[new_var]);
-        item_data->xB8_itemLogicTable = &lbl_803F3100[new_var];
-        goto block_end;
-    }
-    if (ItemID < It_Kind_Old_Kuri) {
-        ItemID -= Pokemon_Tosakinto;
-        new_var = ItemID;
-        ItemID2 = new_var;
-        item_data->xC4_article_data = &(*lbl_804D6D30->unkptr[new_var]);
-        item_data->xB8_itemLogicTable = &lbl_803F23CC[new_var];
-        goto block_end;
-    }
-    ItemID -= It_Kind_Old_Kuri;
-    item_data->xC4_article_data = &(*lbl_804A0F60->unkptr[ItemID]);
-    item_data->xB8_itemLogicTable = &lbl_803F4D20[ItemID];
-    if (item_data->xC4_article_data == NULL) {
-        OSReport("not found zako model data! check ground dat file!\n");
-        __assert("item.c", 0x2AEU, "0");
-    }
-block_end:
+    } else if (item_data->x10_item_kind < Pokemon_Tosakinto) {
+        // Character items
+        int idx = item_data->x10_item_kind - It_Kind_Kuriboh;
+        item_data->xC4_article_data = lbl_804D6D38->unkptr[idx];
+        item_data->xB8_itemLogicTable = &lbl_803F3100[idx];
+    } else if (item_data->x10_item_kind < It_Kind_Old_Kuri) {
+        // Pokemon
+        int idx = item_data->x10_item_kind - Pokemon_Tosakinto;
+        item_data->xC4_article_data = lbl_804D6D30->unkptr[idx];
+        item_data->xB8_itemLogicTable = &lbl_803F23CC[idx];
+    } else {
+        // Stage items
+        int idx = item_data->x10_item_kind - It_Kind_Old_Kuri;
+        item_data->xC4_article_data = lbl_804A0F60->unkptr[idx];
+        item_data->xB8_itemLogicTable = &lbl_803F4D20[idx];
+        if (item_data->xC4_article_data == NULL) {
+            OSReport("not found zako model data! check ground dat file!\n");
+            __assert("item.c", 0x2AE, "0");
+        }
+    }
```

Comment:
The difference of two ITKINDs isn't an ITKIND though, it's an int.

Reply:
Well yeah but these aren't being diff'd. Also, arguably that's not true, based on how HAL uses them... (subtract a fixed value and then compare a range, etc.)

Reply:
When you subtract a fixed value from an enum you don't have the same enum type anymore, you have some kind of index. And they are being diff'd here, they're using that index to access an array of pokemon item tables, or an array of stage item tables, etc. If you treated the result as the same type of enum the result would be wrong and confusing (e.g. it would have the same value as a Capsule item type but it's actually the zero-th index pokemon, i.e. Pokemon_Tosakinto).

Reply:
Oh you're right, I missed the base index subtraction, sorry.